crates/hulk/src/nao/double_buffered_reader.rs

```diff
@@ -0,0 +1,411 @@
+use std::{
+    io::{self, ErrorKind, Read},
+    mem::size_of,
+    os::unix::{io::AsRawFd, prelude::RawFd},
+};
+
+use super::hula::StateStorage;
+
+pub struct DoubleBufferedReader {
+    buffers: [[u8; size_of::<StateStorage>()]; 2],
+    active_buffer_index: usize,
+    incomplete_offset: usize,
+}
+
+impl Default for DoubleBufferedReader {
+    fn default() -> Self {
+        Self {
+            buffers: [[0; size_of::<StateStorage>()]; 2],
+            active_buffer_index: Default::default(),
+            incomplete_offset: Default::default(),
+        }
+    }
+}
+
+impl DoubleBufferedReader {
+    fn swap_buffers(&mut self) {
+        self.incomplete_offset = 0;
+        self.active_buffer_index = (self.active_buffer_index + 1) % 2;
+    }
+
+    pub fn read(
+        &mut self,
+        reader: &mut (impl AsRawFd + Read),
+        mut poll_reader: impl FnMut(RawFd) -> io::Result<()>,
+    ) -> io::Result<()> {
+        let mut buffers_swapped = false;
+        loop {
+            match reader.read(&mut self.buffers[self.active_buffer_index][self.incomplete_offset..])
+            {
+                Ok(number_of_read_bytes) => {
+                    self.incomplete_offset += number_of_read_bytes;
+                    assert!(self.incomplete_offset <= self.buffers[self.active_buffer_index].len());
+                    if self.incomplete_offset == self.buffers[self.active_buffer_index].len() {
+                        self.swap_buffers();
+                        buffers_swapped = true;
+                    }
+                }
+                Err(ref error) if error.kind() == ErrorKind::WouldBlock => {
+                    if buffers_swapped {
+                        return Ok(());
+                    }
+                    poll_reader(reader.as_raw_fd())?;
+                }
+                Err(error) => return Err(error),
+            }
+        }
+    }
+
+    /// Precondition: `read()` has been called once
+    pub fn get_last(&self) -> &StateStorage {
+        let inactive_buffer_index = (self.active_buffer_index + 1) % 2;
+        unsafe { &*(self.buffers[inactive_buffer_index].as_ptr() as *const StateStorage) }
+    }
+}
+
+#[cfg(test)]
```

Comment:
I did not read the tests yet, but think about splitting the tests into "unit-tests" in this module file, and integration tests, which test the API available to the public user.

Reply:
IIRC testing public API is not the separation argument between unit and integration tests. Integration tests are testing a unit together with other units. Since I used traits to only implement the functionality that I need, I would argue that the `DoubleBufferedReader` is a unit and all tests in this section mock all I/O and therefore fall into the unit-test category. Correct me if I'm wrong. :wink:

Reply:
We could also add integration tests but I think this piece of code will be tested enough in using the NAO in everyday life situations.

Reply:
Maybe my current neovim setup just lacks easy folding of parts, but I am always a little distracted by these huge parts of testing code in the implementation. But I think there is not a pretty fix to it. Feel free to resolve when read or start a discussion ;)

Reply:
Well we could move the code into its own file but then we would do it differently than in the other files. Do we want this Präzedenzfall? Do we want this everywhere? I think it is okay to have it near the implementations since it is always at the bottom and noone forces you to scroll that far. Now you resolve again 😉